testnet/stacks-node/src/run_loop/neon.rs

```diff
@@ -173,16 +192,89 @@ impl RunLoop {
                 start_serving_monitoring_metrics(prometheus_bind);
             });
         }
+        
+        let chainstate_path = self.config.get_chainstate_path();
+        let mut unprocessed_block_samples : VecDeque<i64> = VecDeque::new();
+        let max_samples = 30;
+        let max_staging = 10;
+        let delay_heuristic = 60_000;       // 1 minute
+
+        // how often to sync
+        let sync_frequency = 10;
+        let mut sync_deadline = get_epoch_time_secs() + sync_frequency;
 
         loop {
-            burnchain_tip = match burnchain.sync() {
+            let mut total_blocks_seen = 0;
+            let mut total_recent_changes = 0;
+            let do_sync = match self.count_attachable_stacks_blocks(&chainstate_path, max_staging) {
+                Ok(num_available) => {
+                    unprocessed_block_samples.push_back(num_available as i64);
+                    if unprocessed_block_samples.len() > max_samples {
+                        unprocessed_block_samples.pop_front();
+                    }
+
+                    if num_available > 0 || unprocessed_block_samples.len() < max_samples {
+                        // wait for a bit before asking again (up to a minute total)
+                        sleep_ms((delay_heuristic / max_samples) as u64);
+                    }
+
+                    // take first derivative of samples
```

Comment:
First, this derivative sampling should be factored into a separate struct such that there's some simple interface for the sampler that just outputs the current derivative -- combining it with this main loop makes tracking the variable states a little difficult.

Second, I think a node should only be doing this sampling and waiting during startup, _not_ during normal mining operation. If a node was doing this during normal mining operation, it would always wait one minute after receiving any Stacks block before trying to mine.

Reply:
> First, this derivative sampling should be factored into a separate struct such that there's some simple interface for the sampler that just outputs the current derivative -- combining it with this main loop makes tracking the variable states a little difficult.

Can do.

> Second, I think a node should only be doing this sampling and waiting during startup, not during normal mining operation. If a node was doing this during normal mining operation, it would always wait one minute after receiving any Stacks block before trying to mine.

I can make the node infer when it has finished its initial block download, and have it change its mining behavior.

Reply:
Both of these points are now addressed.